ASoC: SOF: ipc4/Intel: Improve pcm delay reporting

ujfalusi opened the pull request:
Hi,
this PR is a new implementation of the closed #4791, providing (I hope) more robust support.

main highlights:
- fix for pause/resume and xrun handling: re-read the start offset from firmware
- callback for reading host accessible LLP has been renamed and
- new host DMA position query callback added
- the delay calculation itself is fixed by using the up-to-date host position, wrapping the LLP with the host boundary and to special case the time while the first frame is still within the DSP processing pipeline

With these patches and an improved audio conformance test (to take the delay into account) we are passing the test quite nicely on hw:0,0, hw:0,31 and DSPless/legacy mode (the later passed before, but indicates that the improved conformance test is not regressing):

command line for the test runs:
./alsa_conformance_test/alsa_conformance_test -P hw:0,X --iterations=20 -B4800 -p4800-d2

hw:0,0
```
----------RUN RESULT----------
number of recorders: 20
number of points: 1214559
step average: 1.496892
step min: 1
step max: 7
rate average: 47999.996679
rate min: 47999.991852
rate max: 48000.001621
rate error average: 0.290328
rate error min: 0.288493
rate error max: 0.291660
number of underrun: 0
number of overrun: 0
```
hw:0,31 (100ms deep buffer)
```
----------RUN RESULT----------
number of recorders: 20
number of points: 582554
step average: 3.001928
step min: 1
step max: 9
rate average: 47999.988525
rate min: 47999.967958
rate max: 48000.006057
rate error average: 0.101782
rate error min: 0.083003
rate error max: 0.115703
number of underrun: 0
number of overrun: 0
```
Note on the steps for the DB case: it is max 9 despite the fact that the DMA jumps 4800 first then 4608 frames at each burst. This is because the test now takes the delay into account which 'corrects' this jumping.

ujfalusi committed:
fixup! ASoC: SOF: Intel: hda: Implement get_stream_position (Linear Link Position)

Rename the function to hda_dsp_get_stream_llp and add a brief function
description.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: Introduce a new callback pair to be used for PCM delay reporting

For delay calculation we need two information:
Number of bytes transferred between the DSP and host memory (ALSA buffer)
Number of frames transferred between the DSP and external device
(link/codec/DMIC/etc).

The reason for the different units (bytes vs frames) on host and dai side
is that the format on the dai side is decided by the firmware and might
not be the same as on the host side, thus the expectation is that the
counter reflects the number of frames.
The kernel know the host side format and in there we have access to the
DMA position which is in bytes.

In a simplified way, the DSP caused delay is the difference between the
two counters.

The existing get_stream_position callback is defined to retrieve the frame
counter on the DAI side but it's name is too generic to be intuitive and
makes it hard to define a callback for the host side.

This patch introduces a new set of callbacks to replace the
get_stream_position and define the host side equivalent:
get_dai_frame_counter
get_host_byte_counter

Subsequent patches will remove the old callback.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: Intel: Set the dai/host get frame/byte counter callbacks

Add implementation for reading the LDP (Linear DMA Position) to be used as
get_host_byte_counter().
The LDP is counting the number of bytes moved between the DSP and host
memory.

Set the get_dai_frame_counter to hda_dsp_get_stream_llp, which is counting
the frames on the link side of the DSP.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: ipc4-pcm: Use the snd_sof_pcm_get_dai_frame_counter() for pcm_delay

Switch to the new callback to retrieve the DAI (link) frame counter.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: Intel: hda-common-ops: Do not set the get_stream_position callback

The get_stream_position has been replaced by get_dai_frame_counter, it
should not be set to allow it to be dropped from core code.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: Intel: Remove the get_stream_position callback

The get_stream_position has been replaced by get_dai_frame_counter and all
related code can be dropped form the core.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: ipc4-pcm: Correct the delay calculation

This patch improves the delay calculation by relying on the
LLP (Linear Link Position) on the DAI side and the
LDP (Linear Data Pointer) on the host side. The LDP provides the same DMA
position as LPIB, but with a linear count instead of a position in the
ALSA ring buffer. The LDP values are provided in bytes and must be
converted to frames. The difference in units means that the host counter
will wrap earlier than the LLP. We need to wrap the LLP at the same
boundary as the host counter.

The ASoC framework relies on separate pointer and delay callback.
Measurement errors can be reduced by processing all the counter values in
the pointer callback. The delay value is stored, and will be reported to
higher levels in the delay callback.

For playback, the firmware provides a stream_start offset to handle
mixing/pause usages, where the DAI might have started earlier than the
PCM device. The delay calculation must be special-cased when the link
counter has not reached the start offset value, i.e. no valid audio has
left the DSP.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: ipc4-pcm: Move struct sof_ipc4_timestamp_info definition locally

The sof_ipc4_timestamp_info is only used by ipc4-pcm.c internally, it
should not be in a generic header implying that it might be used elsewhere.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: ipc4-pcm: Combine the SOF_IPC4_PIPE_PAUSED cases in pcm_trigger

The SNDRV_PCM_TRIGGER_PAUSE_PUSH does not need to be a separate case, it
can be handled along with STOP and SUSPEND

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: ipc4-pcm: Invalidate the stream_start_offset in PAUSED state

When the final state is SOF_IPC4_PIPE_PAUSED, it is possible that the
stream will be restarted (resume or start) in which case we need to update
the offset from the firmware.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

kv2019i commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -884,6 +887,14 @@ static int sof_ipc4_get_stream_start_offset(struct snd_sof_dev *sdev,
 	time_info->stream_start_offset = stream_start_position;
 
+	/*
+	 * Calculate the wrap boundary need to be used for delay callculation
```
typo: callculation

ujfalusi replied:
OK

kv2019i commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -941,16 +960,29 @@ static snd_pcm_sframes_t sof_ipc4_pcm_delay(struct snd_soc_component *component,
+	}
+
+	/* Wrap the dai position at the boundary where the host position wraps */
+	div64_u64_rem(tmp_ptr, time_info->boundary, &tmp_ptr);
```
Rest of this patch looks good and I can follow the explanations, but I fail to understand how the connection between time_info->boundary and substream->runtime->boundary (still used on L989) is made. How do we know host position wraps at "time_info->boundary"? And shouldn't we use time_info->boundary on L989?

ujfalusi replied:
@kv2019i, oh yes, you are right, I need to swap that as well, thanks for finding it!, it would have taken few decades to hit the wrap ;)

kv2019i reviewed (changes requested):
First 9 patches ok, one question on patch 10, please see inline.

ujfalusi commented:
Changes since v1:
- use the correct boundary (time_info->boundary) when the pointers wrap compared to each other.
- fix misspelled word in comment

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -270,6 +270,27 @@ struct snd_sof_dsp_ops {
+	 */
+	u64 (*get_host_byte_counter)(struct snd_sof_dev *sdev,
+				     struct snd_soc_component *component,
+				     struct snd_pcm_substream *substream);
```
nit pick: missing "/* optional */ at the end of line 292 if you want to be consistent with the rest of this header.

ujfalusi replied:
Oh, right, I missed it, I will add it.

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -478,14 +478,13 @@ static int sof_ipc4_pcm_trigger(struct snd_soc_component *component,
 	/* determine the pipeline state */
 	switch (cmd) {
-	case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
-		state = SOF_IPC4_PIPE_PAUSED;
```
need to also remove the break on line 481

ujfalusi replied:
Oh, why the compiler was not complaining? Certainly!

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -21,12 +21,15 @@
  * @dai_copier: the dai copier of the pcm stream
  * @stream_start_offset: reported by fw in memory window
  * @llp_offset: llp offset in memory window
+ * @boundary: wrap boundary should be used for the LLP frame counter
```
Commit message is really odd

"
To be able to calculate the current delay we need to use the DMA position
directly instead of the substream->runtime->status->hw_ptr, which is the
previous hw_pointer and it can be significantly different than where the
hardware currently is.
"

The way I read it is that with this PR we have an accurate delay calculation, but it's not related to the hw_ptr, so if an application does a calculation of DMA position + delay, then it won't be the sample currently playing...

I am struggling to understand the concept, the delay is a relative concept to the DMA position. Finding out the delay independently of the DMA position reported to userspace is a sure way to add uncertainty and measurement problems, no?

ujfalusi replied:
@plbossart , OK, so the issue is two-fold, the use of `substream->runtime->status->hw_ptr,` was not correct as it is not the correct information needed for the delay reporting.

Then to the delay reporting:
when there is no delay (reported) it implies that the frame at the hw_ptr will be on the output at the next sample tick, right?
The delay tells the user space that the frame at the hw_ptr will _not_ be on the output until 'delay' number of frames worth of time passes, right?

It would be ideal if we could ask the firmware to tell us the delay, but it is not doing that, it only gives us two numbers: **stream_start_offset** and the frame counter on the dai side (**LLP**). This is not enough information, but how would the firmware - if it would - gave us a delay number?
When a stream is started it would calculate the **stream_start_offset** which is in essence the number of frame time it will take for the first sample copied to the DSP memory to be on the output.
The **LLP** is the number of frames left the DSP.
On stream start: we copied X frames to the DSP, it will take  Y frame time to reach the output:
when LLP=0 the delay is Y for the first frame, but the host needs to know the delay for the last frame that has been copied to the DSP (the hw_ptr is pointing to the next frame after this last one in the DSP after all), so for the delay that we report to user is: Y+X.
When LLP moves 1 frame, the delay is going to decrease: (Y-1)+X
When LLP moves 1 more frame: (Y-2)+X
...
When LLP moved 100 frames: (Y-100)+X
When LLP == X, that is the point when the first real frame is started to be played out, delay is now X
From this point the delay is: (LLP-Y)+X
and so on, the delay is the difference between the frames entered and left the DSP, adjusted with the stream_start_offset.

Since the firmware is not giving this information, we need to calculate it. We get the LLP and stream_start_offset (Y) from firmware but we need to get the X.
The X is _not_ the DMA pointer in the ALSA buffer, it has nothing to do with ALSA buffer. It is a counter between the host and DSP. Sure, the DMA is counting it but it is rather a counter than a pointer.
We just don't have other means of asking how much data entered or held within the DSP.
Other SoCs, codecs can tell you their FIFO occupation number (which they get by doing the same calculation of what entered and what left).

and why it is not appropriate to use the DMA position in the ALSA buffer?
Our topology tells alsa that the period_size_min is 192 (48 frames in stereo, 16bit - 1ms), the stream_start_offset is about 200 frames.
If application takes three 48 frame periods - which is allowed - then at start the delay is 200+96 frames (our DMA buffer is 2ms on the firmware side), that is hard to represent within the 144 frame long buffer...

The DMA pointer has nothing to do with the delay calculation.
The ALSA buffer has nothing to do with the delay calculation.

What matters:
The time that it takes for a sample to traverse within the DSP (stream_start_offset)
The number of frames left the DSP (LLP)
The number of frames entered the DSP (frame counter on host)

It is an ASoC thing that the pointer and delay calls got separated, but they are executed from the same ALSA pointer call, so yes, there can be some misalignment but that affects _all_ ASoC drivers.

ujfalusi replied:
@plbossart, I think the main issue is that we don't really have a place for this to report, imho.

In essence we would need something somehow use  a counter on the BE and FE side and a start_offset, there is no API to get counter from neither side so we use the 'best' available place where we can do this calculation.

This is in some way a DMA (platform/PCM) caused delay as we stuck the DSP in between the host and link DMA which causes the delay.

plbossart replied:
I am not following your logic @ujfalusi 

My starting point is this: https://elixir.bootlin.com/linux/latest/source/sound/soc/soc-pcm.c#L1220

The soc_pcm_pointer() will get an offset and *IMMEDIATELY* after add a delay.

I don't see the point of rechecking what the host-pointer is, and I don't understand what is calculated by the alsa core. In the ASoC case it looks rather simple: nothing is calculated by the alsa core....

We could get the delay at the same time as the pcm_ pointer, and store it in a cache variable and return it immediately when the snd_soc_pcm_component_delay is used.

the pointer() and delay() callbacks should not be seen as isolated functionality, ideally the pointer and delay should be computed at the same time.

ujfalusi replied:
> I am not following your logic @ujfalusi

@plbossart, OK, I got your point and understand it, but,
 
> My starting point is this: https://elixir.bootlin.com/linux/latest/source/sound/soc/soc-pcm.c#L1220

Yes, I know and for me that whole thing is a bit broken. it is taking the goes through the dais/components and returns the biggest delay reported by the dais/components checked. Imho it should be _adding_ the delays together.
But that is another discussion...
 
> The soc_pcm_pointer() will get an offset and _IMMEDIATELY_ after add a delay.

Yes, and the component delay should be first then the dai delay query, but that is again another discussion.
 
> I don't see the point of rechecking what the host-pointer is, 

We don't check the host-pointer two times, that is the whole point. In `snd_soc_pcm_component_pointer()` we check the host-pointer via DPIB. The pointer is a DMA position within the ALSA buffer.
In the delay callback we calculate the current delay the DSP causes, for which we use two counters and an offset value. DAI counter (LLP) and host counter (LDP) and stream_start_offset. LLP/LDP is counting the number of frames/bytes on the two side of the DSP.

> and I don't understand what is calculated by the alsa core. In the ASoC case it looks rather simple: nothing is calculated by the alsa core....

Under ASoC we always have ALSA doing it's magic:
https://elixir.bootlin.com/linux/latest/source/sound/core/pcm_lib.c#L283
 
> We could get the delay at the same time as the pcm_ pointer, and store it in a cache variable and return it immediately when the snd_soc_pcm_component_delay is used.

In either way you cannot guaranty atomic operation. You still have two calls to get the pointer needed for the pointer and another to get the delay. The host DMA can move in between, it can move between taking the host counter and then the DAI counter or if you reverse them, the DAI counter has moved when you would take the host counter.

If we cache the delay value from taken at SOF pointer time and we have a codec reporting delay then again we have a DSP delay from the past with a current delay on the codec.

It is just that we cannot freeze the system to take the sample, it is moving.

Note: I did experiment similar to this, it was 'off' about the same frequency as the current implementation.

> the pointer() and delay() callbacks should not be seen as isolated functionality, ideally the pointer and delay should be computed at the same time.

For a platform driver, yes, likely it would be better but the DAI, codec, amp can also add delay and there is no way that you can stop everything and read them at the same time.

plbossart replied:
> > I am not following your logic @ujfalusi
> 
> @plbossart, OK, I got your point and understand it, but,
> 
> > My starting point is this: https://elixir.bootlin.com/linux/latest/source/sound/soc/soc-pcm.c#L1220
> 
> Yes, I know and for me that whole thing is a bit broken. it is taking the goes through the dais/components and returns the biggest delay reported by the dais/components checked. Imho it should be _adding_ the delays together. But that is another discussion...

The max() only applies to the case where there are multiple cpu or codec dais. It was selected because there's no good way of determining what the delay should be if multiple dais report different values.

> > I don't see the point of rechecking what the host-pointer is,
> 
> We don't check the host-pointer two times, that is the whole point. In `snd_soc_pcm_component_pointer()` we check the host-pointer via DPIB. The pointer is a DMA position within the ALSA buffer. In the delay callback we calculate the current delay the DSP causes, for which we use two counters and an offset value. DAI counter (LLP) and host counter (LDP) and stream_start_offset. LLP/LDP is counting the number of frames/bytes on the two side of the DSP.

Well you are using DPIB and then LDP. That's two ways of looking at the host pointer.
My question is whether we can use ONE register only.

> > and I don't understand what is calculated by the alsa core. In the ASoC case it looks rather simple: nothing is calculated by the alsa core....
> 
> Under ASoC we always have ALSA doing it's magic: https://elixir.bootlin.com/linux/latest/source/sound/core/pcm_lib.c#L283

Isn't this magic code applied AFTER doing the pointer/delay estimation in the platform driver?

> > We could get the delay at the same time as the pcm_ pointer, and store it in a cache variable and return it immediately when the snd_soc_pcm_component_delay is used.
> 
> In either way you cannot guaranty atomic operation. You still have two calls to get the pointer needed for the pointer and another to get the delay. The host DMA can move in between, it can move between taking the host counter and then the DAI counter or if you reverse them, the DAI counter has moved when you would take the host counter.
> 
> If we cache the delay value from taken at SOF pointer time and we have a codec reporting delay then again we have a DSP delay from the past with a current delay on the codec.
> 
> It is just that we cannot freeze the system to take the sample, it is moving.
> 
> Note: I did experiment similar to this, it was 'off' about the same frequency as the current implementation.
> 
> > the pointer() and delay() callbacks should not be seen as isolated functionality, ideally the pointer and delay should be computed at the same time.
> 
> For a platform driver, yes, likely it would be better but the DAI, codec, amp can also add delay and there is no way that you can stop everything and read them at the same time.

I was only trying to look at the component delay and avoid this DPIB/LDP dual use. Maybe it's a second order problem and maybe it doesn't bring any additional precision in practice, I don't know. It's just very odd to use two reference points instead of one.

ujfalusi replied:
@plbossart, I don't want to use LDP for the pointer. We know that it works CAVS2.5+ but older arch?
The DPIB is convenient as it wraps in buffer.

But, in the core we only get the pointer which is wrapped within the buffer regardless if we derived it from DPIB or LDP. We need a counter.
Where we know the LDP we don't know the LLP (and stream_start_offset),

Either way we need to introduce the same new callbacks and just do the very same thing which we do in the delay callback in a new (we don't have ipc_ops for pointer as it was irrelevant what the DSP does - we cared about the host DMA in buffer). That might get us few usec closer to the pointer time readout of the counter but I don't see the benefit

ujfalusi committed:
ASoC: SOF: Add dsp_max_burst_size_in_ms member to snd_sof_pcm_stream

The dsp_max_burst_size_in_ms can be used to save the length of the maximum
burst size in ms the host DMA will use.

Platform code can place constraint using this to avoid user space
requesting too small ALSA buffer which will result xruns.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: ipc4-topology: Save the DMA maximum burst size for PCMs

When setting up the pcm widget, save the DSP buffer size (in ms) for
platform code to place a constraint on playback.

On playback the DMA will fill the buffer on start and if the period
size is smaller it will immediately overrun.

On capture the DMA will move data in 1ms bursts.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: Intel: hda-pcm: Use dsp_max_burst_size_in_ms to place constraint

If the PCM have the dsp_max_burst_size_in_ms set then place a constraint
to limit the minimum buffer time to avoid xruns caused by DMA bursts
spinning on the ALSA buffer.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi commented:
Changes since v4:
- the leftover 'break;' is removed
- marked the get_host_byte_counter callback as optional
- reworded the last commit  message

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -940,20 +959,33 @@ static snd_pcm_sframes_t sof_ipc4_pcm_delay(struct snd_soc_component *component,
+	 * Otherwise the first audio frames is still within the DSP, we should
+	 * adjust the host side position to accurately report the initial delay
+	 */
+	if (tmp_ptr >= time_info->stream_start_offset) {
```
isn't stream_start_offset only valid for playback?

ujfalusi replied:
Yes, it looks like it is 0 for capture. In that case we report the difference between the host and dai counters w/o compensation.

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -940,20 +959,33 @@ static snd_pcm_sframes_t sof_ipc4_pcm_delay(struct snd_soc_component *component,
 
 	if (head_ptr < tail_ptr)
-		return substream->runtime->boundary - tail_ptr + head_ptr;
+		return time_info->boundary - tail_ptr + head_ptr;
```
does this work for both capture and playback?

ujfalusi replied:
we select the head and tail previously based on direction and the the delay is the difference, if needs to be wrapped, it is the same thing for the two direction.

plbossart reviewed (commented):
other sets of comments.

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -940,20 +959,33 @@ static snd_pcm_sframes_t sof_ipc4_pcm_delay(struct snd_soc_component *component,
+		tmp_ptr -= time_info->stream_start_offset;
+	} else {
+		host_ptr += time_info->stream_start_offset - tmp_ptr;
+		tmp_ptr = 0;
```
having a hard time with both the comments and the code. Why would we adjust the host position?

ujfalusi replied:
I will change the name of the variables, but mainly it is because the tmp and host is in u64 and I want to avoid underflow. The important thing here is the difference between the two counter, that is the delay.

plbossart reviewed (commented):
one more

ujfalusi committed:
ASoC: SOF: sof-pcm: Add pointer callback to sof_ipc_pcm_ops

The IPC specific pointer callback can be used when additional or custom
handling is needed during the pointer calculation, like executing a delay
calculation at the same time to minimize drift between the reported pointer
and the calculated delay.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi commented:
Changes since v5:
- Introduce sof_ipc_pcm_ops.pointer callback (@plbossart FYI)
- Move the delay calculation to the new pointer callback and use the host byte counter to calculate the pointer value
- In the delay callback only return the pre-calculated delay value
- Reword the comments  and rename the variables to indicate that they are counters

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -916,46 +937,108 @@ static snd_pcm_sframes_t sof_ipc4_pcm_delay(struct snd_soc_component *component,
+
+	spcm = snd_sof_find_spcm_dai(component, rtd);
+	if (!spcm)
+		return 0;
```
it's a bit inconsistent to return 0 here and -EOPNOTSUPP in the pointer callback. Maybe it's because we don't test for negative delays, and errors would be caught in the pointer callback anyways?

A comment would help.

ujfalusi replied:
OK,
I needed to use `int` return for the pointer() callback as the pointer is `snd_pcm_uframes_t`, unsinged.
The delay callback is returning the delay _value_.

I'll try to come up with something to clarify it.

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -916,46 +937,108 @@ static snd_pcm_sframes_t sof_ipc4_pcm_delay(struct snd_soc_component *component,
 		ret = sof_ipc4_get_stream_start_offset(sdev, substream, stream, time_info);
 		if (ret < 0)
-			return 0;
+			return -EOPNOTSUPP;
```
should all the changes in error handling be a different patch?

ujfalusi replied:
Unfortunately not, the definition of the pointer callback is that we will ignore fallback to the default pointer code in case -EOPNOTSUPP is returned just like if the callback itself is not set.
If we cannot get things for the delay we shall still report the pointer. We might have real errors to report or other archs will have.

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -21,12 +21,18 @@
  * @llp_offset: llp offset in memory window
+ * @boundary: wrap boundary should be used for the LLP frame counter
+ * @delay: Calculated and stored in pointer callback to be returned as delay
+ *	   value
```
suggested edit
"Calculated and stored in pointer callback. The stored value is returned in the delay callback."

ujfalusi replied:
OK

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -21,12 +21,18 @@
  * @dai_copier: the dai copier of the pcm stream
  * @stream_start_offset: reported by fw in memory window
  * @llp_offset: llp offset in memory window
+ * @boundary: wrap boundary should be used for the LLP frame counter
```
suggested commit message edit

This patch improves the delay calculation by relying on the LLP (Linear Link Position) on the DAI side and the LDP (Linear Data Pointer) on the host side. The LDP provides the same DMA position as LPIB, but with a linear count instead of a position in the ALSA ring buffer. The LDP values are provided in bytes and must be converted to frames. The difference in units means that the host counter will wrap earlier than the LLP. We need to wrap the LLP at the same boundary as the host counter.

The ASoC framework relies on separate pointer and delay callback,. Measurement errors can be reduced by processing all the counter values in the pointer callback. The delay value is stored, and will be reported to higher levels in the delay callback.

For playback, the firmware provides a stream_start offset to handle mixing/pause usages, where the DAI might have started earlier than the PCM device. The delay calculation must be special-cased when the link counter has not reached the start offset value, i.e. no valid audio has left the DSP.

ujfalusi replied:
OK

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -21,12 +21,18 @@
 	u32 llp_offset;
+
+	u64 boundary;
+	snd_pcm_sframes_t delay;
```
add comment along the lines of "negative values shall be ignored"

ujfalusi replied:
The delay is going to be >= 0 when it is calculated, otherwise it remains -EOPNOTSUPP to indicate that the delay is not calculated, which does not make much sense, does it?
I just initialize it to 0 and return it in the delay callback....

plbossart reviewed (commented):
I like this new version @ujfalusi !

I added a couple of comments but this is mostly cosmetic.

ujfalusi committed:
ALSA: hda: Add pplcllpl/u members to hdac_ext_stream

The pplcllpl/u can be used to save the Link Connection Linear Link
Position register value to be used for compensation of the LLP register
value in case the counter is not reset (after pause/resume or
stop/start without closing the stream).

The LLP can be used along with PPHCLDP to calculate delay caused by the DSP
processing for HDA links.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: Intel: hda: Compensate LLP in case it is not reset

During pause/reset or stop/start the LLP counter is not reset, which will
result broken delay reporting.

Read the LLP value on STOP/PAUSE trigger and use it in LLP reading to
normalize the LLP from the register.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi commented:
Changes since v6:
- commit message and comments added to clarify some details
- the delay in time_info is initialized to 0, so no magic needed in delay() callback - also documented in comments

ujfalusi commented:
@plbossart, @kv2019i, @ranj063, we only have one major remaining issue with the delay: pause/resume or xrun or stop/start w/o stream close on HDA link (when the host register is used for the LLP):
The LLP is _not_ reset to 0 while the host counter is. I could not find a way to reset the LLP back to 0 (not even with playing with RSM bit), so I ended up with these commit:
https://github.com/ujfalusi/sof-linux/commit/8d0b8baabe0698ed3d2f6124990d438f2f138418
https://github.com/ujfalusi/sof-linux/commit/b2decbf8fee9f18fb65cc60e8f994718416cdb47

with these + plus the constraint things look pretty good on SDW and HDA links and with various media players.

plbossart commented:
> @plbossart, @kv2019i, @ranj063, we only have one major remaining issue with the delay: pause/resume or xrun or stop/start w/o stream close on HDA link (when the host register is used for the LLP): The LLP is _not_ reset to 0 while the host counter is. 

for playback doesn't the start_offset take care of the issue? In theory the LLP can keep going and the firmware would compensate, no?

ujfalusi commented:
> > @plbossart, @kv2019i, @ranj063, we only have one major remaining issue with the delay: pause/resume or xrun or stop/start w/o stream close on HDA link (when the host register is used for the LLP): The LLP is _not_ reset to 0 while the host counter is.
> 
> for playback doesn't the start_offset take care of the issue? In theory the LLP can keep going and the firmware would compensate, no?

Well, in theory yes, the code is there.
In reality the link DMA is always stopped in firmware and the start_offset is always calculated assuming that the link counter is starting from 0 (there is never a `COMP_TRIGGER_RELEASE`, only `COMP_TRIGGER_START`). The non HDA LLP is indeed reset to 0 but for some reason the HDA LLP (host accessible) is not. Firmware has no access to this LLP so it needs to be compensated on host side.

ujfalusi commented:
Changes since v7:
- Merged the patches from #4858 
  -  The constraint is placed on the buffer time instead of period time. (2x host DMA buffer size)
- Added patches to fix the pause/resume, xrun or stop/start w/o stream close/open issue on HDA links
  - we need to use the LLP value which was at stop as compensation since the host accessible LLP is not reset

kv2019i reviewed (commented):
Tested this on MTL system with SDW codecs and works now great with seeks (tested multiple applications), both regular PCMs and deepbuffer configuration. Pauses continue be a problem, easily triggered with mplayer/mpv.

ujfalusi commented:
> Tested this on MTL system with SDW codecs and works now great with seeks (tested multiple applications), both regular PCMs and deepbuffer configuration. Pauses continue be a problem, easily triggered with mplayer/mpv.

Yes, it can be also tested with `aplay -i`, basically the firmware starts to report drifting and incorrect stream_start_offset and stream_end_offset values and that blows up the kernel reporting.
So far I was not able to find any way to make sense of the situation, It cannot be fixed in kernel.

A) SNDRV_PCM_INFO_PAUSE
B) delay reporting

We cannot have A & B at the moment, it is either A or B.
PAUSE causes t he delay to be broken if used.

plbossart commented on sound/soc/sof/intel/hda-pcm.c:
```diff
@@ -263,6 +263,25 @@ int hda_dsp_pcm_open(struct snd_sof_dev *sdev,
+		snd_pcm_hw_constraint_minmax(substream->runtime,
+			SNDRV_PCM_HW_PARAM_BUFFER_TIME,
+			spcm->stream[substream->stream].dsp_buffer_time_ms * USEC_PER_MSEC * 2,
+			UINT_MAX);
```
By default 

````c
sps->dsp_buffer_time_ms = SOF_IPC4_MIN_DMA_BUFFER_SIZE;
````
so this forces the ALSA buffer to be a minimum of 4ms.

Also this is a bit ironic that the race happens for large buffers, IIRC the issue was an xrun with small periods with mplayer? If this is only for large buffers, could we enforce this constraint when 

````spcm->stream[substream->stream].dsp_buffer_time_ms > SOF_IPC4_MIN_DMA_BUFFER_SIZE;````

Bear with me, reviewing on Friday afternoon...

ujfalusi replied:
The xrun happens with large buffers because application rarely if ever took small (< 4ms) buffers.

plbossart replied:
so in practice could we not make this constraint conditional, e.g.

````c
if ( spcm->stream[substream->stream].dsp_buffer_time_ms > SOF_IPC4_MIN_DMA_BUFFER_SIZE) {
    snd_pcm_hw_constraint_minmax(substream->runtime,
			SNDRV_PCM_HW_PARAM_BUFFER_TIME,
			spcm->stream[substream->stream].dsp_buffer_time_ms * USEC_PER_MSEC * 2,
			UINT_MAX);
}

ujfalusi replied:
> so in practice could we not make this constraint conditional, e.g.
> 
> ```c
> if ( spcm->stream[substream->stream].dsp_buffer_time_ms > SOF_IPC4_MIN_DMA_BUFFER_SIZE) {
>     snd_pcm_hw_constraint_minmax(substream->runtime,
> 			SNDRV_PCM_HW_PARAM_BUFFER_TIME,
> 			spcm->stream[substream->stream].dsp_buffer_time_ms * USEC_PER_MSEC * 2,
> 			UINT_MAX);
> }
> ```

No, the ALSA buffer must be constrained at all times, if the PCM uses the default 2ms buffer then the first DMA burst will copy 2ms data, the buffer should be double of that.

While on capture the ALSA buffer size should be at least 2ms, the DMA copies 1ms chunks from DSP to host, this I left out, but I can add.

plbossart reviewed (commented):
Looks good, only a comment on the last patch @ujfalusi

marc-hb commented:
SOFCI TEST

ujfalusi commented:
Changes since v8:
- Moved the constraint patches earlier
- renamed the `dsp_buffer_time_ms` to `dsp_max_burst_size_in_ms` and set it to 1ms for capture

kv2019i reviewed (approved):
Thanks @ujfalusi , looks good now.

plbossart commented:
Feel free to merge @ujfalusi, I can't recall if we need to disable pause as well or if your last patch corrects the problem.

ujfalusi commented:
> Feel free to merge @ujfalusi, I can't recall if we need to disable pause as well or if your last patch corrects the problem.

I would try to fix the firmware or at least make it less broken. It would be a major pain to deal with the disable and then re-enabling of the pause.
